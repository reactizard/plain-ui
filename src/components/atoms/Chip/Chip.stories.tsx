import { Meta, StoryObj } from "@storybook/react"
import { Chip } from "./Chip"

const meta: Meta<typeof Chip> = {
  title: "Atoms/Chip",
  component: Chip,
  tags: ["autodocs"],
}

export default meta
type Story = StoryObj<typeof Chip>

export const Primary: Story = {
  args: {
    text: "Chip",
    color: "primary",
    variant: "filled",
  },
}
