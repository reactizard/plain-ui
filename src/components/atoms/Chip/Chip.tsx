import React, { ButtonHTMLAttributes } from "react"
import useDisclosure from "../../../hooks/useDisclosure"
import Times from "../../../icons/Times"
import { Colors, Variant } from "../../../theme/constants"
import { getClasses } from "../../../theme/functions/getClasses"
import { getStyles } from "./utils/styles"

interface ChipProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  text: string
  variant?: Exclude<Variant, "light">
  color?: Colors
  disabled?: boolean
}

export const Chip: React.FC<ChipProps> = React.forwardRef<
  HTMLButtonElement,
  ChipProps
>(
  (
    {
      text,
      variant = "filled",
      color = "primary",
      disabled,
      ...rest
    }: ChipProps,
    ref
  ) => {
    const [opened, { toggle }] = useDisclosure(true)
    const classes = getStyles({
      variant: variant,
      color: color,
      disabled: disabled,
    })

    const { fill } = getClasses({ color })

    return opened ? (
      <button className={classes} disabled={disabled} ref={ref} {...rest}>
        {text}
        <Times
          customfill={{ color: fill, variant }}
          onClick={() => (!disabled ? toggle() : null)}
        />
      </button>
    ) : null
  }
) as any

export default Chip
