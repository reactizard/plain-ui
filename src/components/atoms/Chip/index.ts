export { default as useDisclosure } from "../../../hooks/useDisclosure"
export { default } from "./Chip"
