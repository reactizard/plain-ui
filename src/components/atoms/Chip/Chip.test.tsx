import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { render } from '@testing-library/react';

import Chip from './Chip';

describe('Chip component', () => {
  test('renders properly', () => {
    const { getByText } = render(<Chip text="Test Chip" />);
    const chipText = getByText('Test Chip');
    expect(chipText).toBeInTheDocument();
  });

  test('shows same UI as expected', () => {
    const { container } = render(<Chip text="Test Chip" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('changing props reflects on the UI', () => {
    const { getByText, rerender } = render(
      <Chip text="Test Chip" variant="outlined" color="primary" disabled />
    );
    const chipText = getByText('Test Chip');

    expect(chipText).toBeInTheDocument();
    expect(chipText).not.toHaveClass('bg-primary-500');
    expect(chipText).toHaveClass('border-primary-700');
    expect(chipText).toHaveClass('text-primary-700');
    expect(chipText).toBeDisabled();

    // Re render with updated props
    rerender(<Chip text="Updated Chip" variant="filled" color="primary" />);

    expect(chipText).toHaveTextContent('Updated Chip');
    expect(chipText).toHaveClass('bg-primary-500');
    expect(chipText).toHaveClass('text-white', 'bg-primary-500');
    expect(chipText).not.toBeDisabled();
  });
});
