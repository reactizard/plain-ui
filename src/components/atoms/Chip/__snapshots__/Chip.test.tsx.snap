// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip component shows same UI as expected 1`] = `
<button
  class="w-fit flex items-center rounded-[4px] gap-[4px] px-[8px] py-[2px] text-smSemiBold bg-primary-500 hover:bg-primary-600 text-gray-100"
>
  Test Chip
</button>
`;
